src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -836,7 +836,9 @@ private void ImportAddressOfVar(int index, bool argument)
 
         private void ImportDup()
         {
-            _stack.Push(_stack.Peek().Duplicate());
+            var entry = _stack.Pop();
+            _stack.Push(entry.Duplicate(_builder));
```

Comment:
I believe this and the line below are switched. From [here](https://msdn.microsoft.com/en-us/library/system.reflection.emit.opcodes.dup(v=vs.110).aspx):
1. value is pushed onto the stack.
2. value is popped off of the stack for duplication.
3. value is pushed back onto the stack.
4. A duplicate value is pushed onto the stack.

Reply:
Prior to the change in the order of duplicate I was still seeing the original value get updated at the wrong time. I wonder if that means both entries that get pushed on the stack need to have been loaded, or if that means there is a more fundamental problem with using LoadExpressionEntry.

Reply:
Loading both does seem reasonable. I think the reason we generally defer loading is just to avoid unnecessary casting, but maybe dup forces it.

Reply:
We defer loading because llvm doesn't let you cast arbitrary values like you can with pointers. There is still a bit of code in the casting logic to deal with creating Temps in the cases where it's needed for casting an already loaded value, but it's really nasty and might have edge case behaviors if you call dup on a structure value then update a member of what should be the original. Before I go searching the tests for it, do you happen to know where this sort of thing is tested in the larger product.

Reply:
It looks like CoreCLR has a number of dup tests in https://github.com/dotnet/coreclr/tree/32f0f9721afb584b4a14d69135bea7ddc129f755/tests/src/JIT/IL_Conformance/Old/Conformance_Base and other usage in various IL tests (take a look at https://github.com/dotnet/corert/blob/master/tests/src/Simple/HelloWasm/CpObj.il for an idea of how to add an IL test to HelloWasm). The search at https://github.com/dotnet/coreclr/search?p=2&q=dup&type=&utf8=%E2%9C%93 is showing some other dup usage in tests as well.

Reply:
Unfortunately those IL conformance tests don't really do much in the way of testing. They just validate that the result of dup equals itself.

Reply:
Roslyn is pretty aggressive with `dup` these days and it has triggered a couple of jit bugs.

You might look at dotnet/coreclr#9945 or dotnet/coreclr#14784 for fairly simple test cases, if you don't feel like writing your own.

Reply:
Thanks, I grabbed one of those and made one of my own.